[Remote Inspection] Add SPI to make it possible to inspect content under targeted elements

whsieh opened the pull request:
#### 10f56d8c541aa03898e9f1120e687d16d604d2a4
<pre>
[Remote Inspection] Add SPI to make it possible to inspect content under targeted elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=271475">https://bugs.webkit.org/show_bug.cgi?id=271475</a>

Reviewed by Aditya Keerthi.

Introduce `-_adjustVisibilityForTargetedElements:completionHandler:`, and hook it up to the newly-
landed visibility adjustment infrastructure in WebCore. See below for more details.

* LayoutTests/TestExpectations:
* LayoutTests/fast/element-targeting/basic-element-targeting-expected.html: Added.
* LayoutTests/fast/element-targeting/basic-element-targeting.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:

Add a test directory (and basic test) to exercise this new SPI.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestRenderedTextForSelector):
(window.UIHelper.adjustVisibilityForFrontmostTarget):
(window.UIHelper):

Add a testing hook that allows layout tests to exercise element targeting.

* Source/WebCore/page/ElementTargeting.cpp:
(WebCore::setNeedsVisibilityAdjustmentRecursive):
(WebCore::adjustVisibilityForTargetedElements):
* Source/WebCore/page/ElementTargeting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _adjustVisibilityForTargetedElements:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):

Add plumbing through `WebPageProxy` and `WebPage`, down into WebCore.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::adjustVisibilityForTargetedElements):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::adjustVisibilityForFrontmostTarget):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::adjustVisibilityForFrontmostTarget):

Canonical link: <a href="https://commits.webkit.org/276573@main">https://commits.webkit.org/276573@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/108c94474e81588669fc2ef4a44a4baac4605288

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44993 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47492 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47651 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41000 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47299 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/28231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21500 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36926 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38755 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18014 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18577 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39881 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3041 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41261 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49325 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/19966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16510 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21281 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38975 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42702 "Found 1 new API test failure: /WebKitGTK/TestLoaderClient:/webkit/WebKitWebView/load-twice-and-reload (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10016 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/page/ElementTargeting.cpp:
```diff
@@ -323,4 +326,45 @@ Vector<TargetedElementInfo> findTargetedElements(Page& page, TargetedElementRequ
+
+bool adjustVisibilityForTargetedElements(const Vector<Ref<Element>>& elements)
+{
+    bool changed = false;
+    for (auto& element : elements) {
+        if (setNeedsVisibilityAdjustmentRecursive(element))
+            changed = true;
+    }
+    return changed;
```
```
return WTF::anyOf(elements, [&] (auto& element) {
    return setNeedsVisibilityAdjustmentRecursive(element);
});
```

whsieh replied:
I think that's actually not _quite_ what I want (since I want to continue calling `setNeedsVisibilityAdjustmentRecursive` on the full list of elements, even if `changed` already becomes true). I'll leave this as-is (unless there's another more elegant way to write this?)

(Edit: to clarify, `WTF::anyOf` will bail early when the lambda returns `true` on the first element)

szewai replied:
ah got it; sounds good!

pxlcoder commented on Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
```diff
@@ -353,4 +355,28 @@ - (void)paste:(id)sender;
+        }
+
+        RetainPtr<_WKTargetedElementInfo> frontTarget = elements.firstObject;
+        [webView() _adjustVisibilityForTargetedElements:@[frontTarget.get()] completionHandler:[callbackID, frontTarget, this] (BOOL success) {
```
Nit:
```suggestion
        [webView() _adjustVisibilityForTargetedElements:@[ frontTarget.get() ] completionHandler:[callbackID, frontTarget, this] (BOOL success) {
```

whsieh replied:
Ah, I had that originally, but the style checker (erroneously) complained about extra spaces around the "capture group" 😓. I'll add the spaces back in (and at some point, we should probably teach the style checker about `@[]`…)

Thanks for the review!

whsieh committed:
[Remote Inspection] Add SPI to make it possible to inspect content under targeted elements
https://bugs.webkit.org/show_bug.cgi?id=271475

Reviewed by Aditya Keerthi.

Introduce `-_adjustVisibilityForTargetedElements:completionHandler:`, and hook it up to the newly-
landed visibility adjustment infrastructure in WebCore. See below for more details.

* LayoutTests/TestExpectations:
* LayoutTests/fast/element-targeting/basic-element-targeting-expected.html: Added.
* LayoutTests/fast/element-targeting/basic-element-targeting.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:

Add a test directory (and basic test) to exercise this new SPI.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestRenderedTextForSelector):
(window.UIHelper.adjustVisibilityForFrontmostTarget):
(window.UIHelper):

Add a testing hook that allows layout tests to exercise element targeting.

* Source/WebCore/page/ElementTargeting.cpp:
(WebCore::setNeedsVisibilityAdjustmentRecursive):
(WebCore::adjustVisibilityForTargetedElements):
* Source/WebCore/page/ElementTargeting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _adjustVisibilityForTargetedElements:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):

Add plumbing through `WebPageProxy` and `WebPage`, down into WebCore.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::adjustVisibilityForTargetedElements):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::adjustVisibilityForFrontmostTarget):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::adjustVisibilityForFrontmostTarget):

Canonical link: https://commits.webkit.org/276573@main
